File: typescript/src/challenges/quiz.ts
/**
 * Quiz challenge for HIGH-risk actions.
 *
 * Generates 1-3 comprehension questions derived from the actual action
 * parameters so the operator must demonstrate they understand what is
 * about to happen before approving.
 */

import * as path from "node:path";

import {
  type ActionContext,
  type ChallengeProtocol,
  type ChallengeResult,
  ChallengeType,
  type RiskAssessment,
  createChallengeResult,
  describeAction,
} from "../types.js";

// ---------------------------------------------------------------------------
// Question model
// ---------------------------------------------------------------------------

/**
 * A single quiz question presented to the operator.
 */
export interface Question {
  /** The question text shown to the operator. */
  text: string;

  /** The canonical correct answer (case-insensitive comparison). */
  correctAnswer: string;

  /**
   * null for free-text input, or a list of choices for multiple-choice
   * (the correct answer must appear in the list).
   */
  options: string[] | null;
}

// ---------------------------------------------------------------------------
// Helpers for question generation
// ---------------------------------------------------------------------------

const PATH_RE =
  /(?:\/(?:[a-zA-Z0-9._\-]+\/)*[a-zA-Z0-9._\-]+)|(?:[A-Z]:\\(?:[^\\\s]+\\)*[^\\\s]+)|(?:(?:\.\.\/|\.\/)(?:[a-zA-Z0-9._\-]+\/)*[a-zA-Z0-9._\-]+)/g;

const SQL_TABLE_RE =
  /(?:FROM|INTO|UPDATE|TABLE|JOIN)\s+[`"']?(\w+)[`"']?/gi;

const NUMERIC_RE = /^-?\d+(?:\.\d+)?$/;

interface NamedValue {
  name: string | null;
  value: unknown;
}

function flattenArgs(ctx: ActionContext): NamedValue[] {
  const items: NamedValue[] = ctx.args.map((a) => ({
    name: null,
    value: a,
  }));
  for (const [key, val] of Object.entries(ctx.kwargs)) {
    items.push({ name: key, value: val });
  }
  return items;
}

function findPaths(values: NamedValue[]): string[] {
  const paths: string[] = [];
  for (const { value } of values) {
    if (typeof value === "string") {
      PATH_RE.lastIndex = 0;
      let match: RegExpExecArray | null;
      while ((match = PATH_RE.exec(value)) !== null) {
        paths.push(match[0]);
      }
    }
  }
  return paths;
}

function findNumbers(values: NamedValue[]): Array<{ name: string | null; numStr: string }> {
  const results: Array<{ name: string | null; numStr: string }> = [];
  for (const { name, value } of values) {
    if (typeof value === "number" && !isNaN(value)) {
      results.push({ name, numStr: String(value) });
    } else if (typeof value === "string" && NUMERIC_RE.test(value)) {
      results.push({ name, numStr: value });
    }
  }
  return results;
}

function findSqlTables(values: NamedValue[]): string[] {
  const tables: string[] = [];
  const seen = new Set<string>();
  for (const { value } of values) {
    if (typeof value === "string") {
      SQL_TABLE_RE.lastIndex = 0;
      let match: RegExpExecArray | null;
      while ((match = SQL_TABLE_RE.exec(value)) !== null) {
        const table = match[1];
        if (!seen.has(table)) {
          seen.add(table);
          tables.push(table);
        }
      }
    }
  }
  return tables;
}

function splitFunctionName(name: string): string[] {
  // Split on underscores first
  const parts = name.replace(/-/g, "_").split("_");
  // Then split camelCase/PascalCase within each part
  const words: string[] = [];
  for (const part of parts) {
    const tokens = part.replace(/([a-z])([A-Z])/g, "$1 $2").split(/\s+/);
    for (const t of tokens) {
      if (t) words.push(t.toLowerCase());
    }
  }
  return words;
}

function shuffleArray<T>(arr: T[]): T[] {
  const result = [...arr];
  for (let i = result.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [result[i], result[j]] = [result[j], result[i]];
  }
  return result;
}

// ---------------------------------------------------------------------------
// QuizChallenge options
// ---------------------------------------------------------------------------

/**
 * Options for configuring the QuizChallenge.
 */
export interface QuizChallengeOptions {
  /** Upper bound on the number of questions to generate (1-3). Default 3. */
  maxQuestions?: number;

  /** Minimum correct answers required to pass. Default 1. */
  minCorrect?: number;

  /** Minimum wall-clock seconds before the first question is shown. Default 10. */
  minReviewSeconds?: number;

  /** Custom input function. Signature: (prompt: string) => Promise<string>. */
  promptFn?: (prompt: string) => Promise<string>;

  /** Custom output function. */
  printFn?: (message: string) => void;
}

// ---------------------------------------------------------------------------
// QuizChallenge
// ---------------------------------------------------------------------------

/**
 * Comprehension quiz for HIGH-risk actions.
 *
 * Analyses the ActionContext to programmatically generate 1-3
 * multiple-choice or fill-in-the-blank questions about the pending action.
 * The operator must answer at least minCorrect questions correctly to pass.
 */
export class QuizChallenge implements ChallengeProtocol {
  readonly maxQuestions: number;
  readonly minCorrect: number;
  readonly minReviewSeconds: number;
  private readonly _promptFn: (prompt: string) => Promise<string>;
  private readonly _printFn: (message: string) => void;

  constructor(options: QuizChallengeOptions = {}) {
    this.maxQuestions = Math.max(1, Math.min(options.maxQuestions ?? 3, 3));
    this.minCorrect = Math.max(1, options.minCorrect ?? 1);
    this.minReviewSeconds = options.minReviewSeconds ?? 10.0;
    this._promptFn = options.promptFn ?? defaultPrompt;
    this._printFn = options.printFn ?? defaultPrint;
  }

  get challengeType(): ChallengeType {
    return ChallengeType.QUIZ;
  }

  // -- question generation -----------------------------------------------

  /**
   * Build a list of questions from the action context.
   *
   * Strategy priority:
   * 1. File-path arguments   -> "What directory / file will be affected?"
   * 2. Numeric arguments     -> "What is the value of <param>?"
   * 3. SQL-like strings      -> "Which table will be affected?"
   * 4. Fallback              -> "What will this function do?" (from name)
   */
  generateQuestions(
    ctx: ActionContext,
    _risk: RiskAssessment
  ): Question[] {
    const flat = flattenArgs(ctx);
    const questions: Question[] = [];

    // Strategy 1: file paths
    const paths = findPaths(flat);
    if (paths.length > 0) {
      const p = paths[0];
      const dirName = path.dirname(p) || "/";
      const baseName = path.basename(p);

      // Ask about the directory
      let wrongDirs = [
        path.dirname(path.dirname(p)) || "/",
        "/tmp",
        "/var/log",
      ];
      // Filter out correct answer and dedupe
      wrongDirs = [...new Set(wrongDirs.filter((d) => d !== dirName))];
      const dirOptions = shuffleArray([dirName, ...wrongDirs.slice(0, 3)]);

      questions.push({
        text: "What directory will be affected by this action?",
        correctAnswer: dirName,
        options: dirOptions,
      });

      // Optionally ask about the specific file
      if (questions.length < this.maxQuestions && baseName) {
        questions.push({
          text: "Which file will this action operate on?",
          correctAnswer: baseName,
          options: null, // free text
        });
      }
    }

    // Strategy 2: numeric arguments
    const numbers = findNumbers(flat);
    if (numbers.length > 0 && questions.length < this.maxQuestions) {
      const { name, numStr } = numbers[0];
      const label = name ? `parameter '${name}'` : "this action";

      let wrong: string[];
      try {
        const numVal = parseFloat(numStr);
        if (Number.isInteger(numVal)) {
          wrong = [
            String(numVal * 2),
            String(numVal + 10),
            String(Math.max(0, numVal - 1)),
          ];
        } else {
          wrong = [
            (numVal * 2).toFixed(2),
            (numVal + 10).toFixed(2),
            Math.max(0, numVal - 1).toFixed(2),
          ];
        }
      } catch {
        wrong = ["0", "100", "42"];
      }

      wrong = [...new Set(wrong.filter((w) => w !== numStr))].slice(0, 3);
      const numOptions = shuffleArray([numStr, ...wrong]);

      questions.push({
        text: `What is the numeric value for ${label}?`,
        correctAnswer: numStr,
        options: numOptions,
      });
    }

    // Strategy 3: SQL table names
    const tables = findSqlTables(flat);
    if (tables.length > 0 && questions.length < this.maxQuestions) {
      const table = tables[0];
      let wrongTables = [
        "users",
        "logs",
        "tmp_data",
        "sessions",
        "config",
      ];
      wrongTables = wrongTables
        .filter((t) => t.toLowerCase() !== table.toLowerCase())
        .slice(0, 3);
      const tableOptions = shuffleArray([table, ...wrongTables]);

      questions.push({
        text: "Which database table will be affected?",
        correctAnswer: table,
        options: tableOptions,
      });
    }

    // Strategy 4 (fallback): function-name comprehension
    if (questions.length === 0) {
      const words = splitFunctionName(ctx.functionName);
      const actionVerb = words[0] ?? "perform";
      const target =
        words.length > 1 ? words.slice(1).join(" ") : "an operation";
      const correct = `${actionVerb} ${target}`;

      let wrong = [
        `read ${target}`,
        `list ${target}`,
        `validate ${target}`,
      ];
      wrong = wrong.filter((w) => w !== correct).slice(0, 3);
      const fallbackOptions = shuffleArray([correct, ...wrong]);

      questions.push({
        text: `What will the function '${ctx.functionName}' do?`,
        correctAnswer: correct,
        options: fallbackOptions,
      });
    }

    return questions.slice(0, this.maxQuestions);
  }

  // -- presentation -------------------------------------------------------

  /**
   * Present the quiz to the operator and collect answers.
   */
  async present(
    ctx: ActionContext,
    risk: RiskAssessment
  ): Promise<ChallengeResult> {
    const start = performance.now();
    const print = this._printFn;
    const prompt = this._promptFn;

    // -- render action summary ------------------------------------------
    const separator = "=".repeat(60);
    print(`\n${separator}`);
    print("  QUIZ CHALLENGE  --  HIGH RISK ACTION");
    print(separator);
    print(`  Action: ${ctx.functionName}`);
    print(
      `  Risk:   ${risk.level.toUpperCase()} (${risk.score.toFixed(2)})`
    );
    if (ctx.functionDoc) {
      print(`  Desc:   ${ctx.functionDoc}`);
    }
    print(`  Call:   ${describeAction(ctx)}`);
    print(separator);

    // -- enforce minimum review time ------------------------------------
    const elapsed = (performance.now() - start) / 1000;
    if (elapsed < this.minReviewSeconds) {
      const remaining = this.minReviewSeconds - elapsed;
      print(
        `  [Review the action for ${Math.round(remaining)}s before answering...]`
      );
      await sleep(remaining);
    }

    // -- generate and ask questions -------------------------------------
    const questions = this.generateQuestions(ctx, risk);
    let correctCount = 0;

    for (let idx = 0; idx < questions.length; idx++) {
      const question = questions[idx];
      print(
        `\n  Question ${idx + 1}/${questions.length}: ${question.text}`
      );

      let answer: string;
      if (question.options) {
        for (let letterIdx = 0; letterIdx < question.options.length; letterIdx++) {
          const letter = String.fromCharCode(65 + letterIdx); // A, B, C, ...
          print(`    ${letter}) ${question.options[letterIdx]}`);
        }
        const raw = await prompt("  Your answer (letter or value): ");
        answer = raw.trim();

        // Accept either the letter label or the literal value
        if (
          answer.length === 1 &&
          answer.toUpperCase() >= "A" &&
          answer.toUpperCase() <= "Z"
        ) {
          const choiceIdx = answer.toUpperCase().charCodeAt(0) - 65;
          if (
            choiceIdx >= 0 &&
            choiceIdx < question.options.length
          ) {
            answer = question.options[choiceIdx];
          }
        }
      } else {
        const raw = await prompt("  Your answer: ");
        answer = raw.trim();
      }

      if (answer.toLowerCase() === question.correctAnswer.toLowerCase()) {
        correctCount++;
        print("  Correct.");
      } else {
        print(
          `  Incorrect. (expected: ${question.correctAnswer})`
        );
      }
    }

    const totalElapsed = (performance.now() - start) / 1000;
    const passed = correctCount >= this.minCorrect;

    print(
      `\n  Result: ${correctCount}/${questions.length} correct -- ${passed ? "PASSED" : "FAILED"}`
    );

    return createChallengeResult({
      passed,
      challengeType: this.challengeType,
      responseTimeSeconds: totalElapsed,
      questionsAsked: questions.length,
      questionsCorrect: correctCount,
      details: {
        questions: questions.map((q) => ({
          text: q.text,
          correctAnswer: q.correctAnswer,
          options: q.options,
        })),
      },
    });
  }
}

// ---------------------------------------------------------------------------
// Default I/O helpers
// ---------------------------------------------------------------------------

function defaultPrint(message: string): void {
  // eslint-disable-next-line no-console
  console.log(message);
}

async function defaultPrompt(message: string): Promise<string> {
  const { createInterface } = await import("node:readline");
  const rl = createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  return new Promise<string>((resolve) => {
    rl.question(message, (answer) => {
      rl.close();
      resolve(answer);
    });
  });
}

function sleep(seconds: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, seconds * 1000));
}

File: python/src/attesta/challenges/quiz.py
"""Quiz challenge for HIGH-risk actions.

Generates 1-3 comprehension questions derived from the actual action
parameters so the operator must demonstrate they understand *what* is
about to happen before approving.
"""

from __future__ import annotations

import asyncio
import os
import random
import re
import time
from dataclasses import dataclass
from typing import Any

from attesta.core.types import (
    ActionContext,
    ChallengeResult,
    ChallengeType,
    RiskAssessment,
)

# ---------------------------------------------------------------------------
# Question model
# ---------------------------------------------------------------------------


@dataclass
class Question:
    """A single quiz question presented to the operator.

    Parameters
    ----------
    text:
        The question text shown to the operator.
    correct_answer:
        The canonical correct answer (case-insensitive comparison).
    options:
        ``None`` for free-text input, or a list of choices for
        multiple-choice (the correct answer **must** appear in the list).
    """

    text: str
    correct_answer: str
    options: list[str] | None = None  # None = free text, list = multiple choice


# ---------------------------------------------------------------------------
# Helpers for question generation
# ---------------------------------------------------------------------------

_PATH_RE = re.compile(
    r"""
    (?:                          # Absolute paths
        /(?:[a-zA-Z0-9._\-]+/)*[a-zA-Z0-9._\-]+
    )
    |
    (?:                          # Windows-style paths
        [A-Z]:\\(?:[^\\\s]+\\)*[^\\\s]+
    )
    |
    (?:                          # Relative paths with at least one separator
        (?:\.\./|\./)(?:[a-zA-Z0-9._\-]+/)*[a-zA-Z0-9._\-]+
    )
    """,
    re.VERBOSE,
)

_SQL_TABLE_RE = re.compile(
    r"""
    (?:FROM|INTO|UPDATE|TABLE|JOIN)\s+
    [`"']?(\w+)[`"']?
    """,
    re.IGNORECASE | re.VERBOSE,
)

_NUMERIC_RE = re.compile(r"^-?\d+(?:\.\d+)?$")


def _flatten_args(ctx: ActionContext) -> list[tuple[str | None, Any]]:
    """Return a flat list of ``(name_or_none, value)`` from args + kwargs."""
    items: list[tuple[str | None, Any]] = [(None, a) for a in ctx.args]
    items.extend(ctx.kwargs.items())
    return items


def _find_paths(values: list[tuple[str | None, Any]]) -> list[str]:
    """Extract file-system paths from argument values."""
    paths: list[str] = []
    for _, val in values:
        if isinstance(val, str):
            paths.extend(_PATH_RE.findall(val))
    return paths


def _find_numbers(values: list[tuple[str | None, Any]]) -> list[tuple[str | None, str]]:
    """Return ``(param_name, numeric_str)`` for numeric args."""
    results: list[tuple[str | None, str]] = []
    for name, val in values:
        str_val = str(val)
        if isinstance(val, (int, float)) and not isinstance(val, bool):
            results.append((name, str_val))
        elif isinstance(val, str) and _NUMERIC_RE.match(val):
            results.append((name, val))
    return results


def _find_sql_tables(values: list[tuple[str | None, Any]]) -> list[str]:
    """Extract SQL table names from string arguments."""
    tables: list[str] = []
    for _, val in values:
        if isinstance(val, str):
            tables.extend(_SQL_TABLE_RE.findall(val))
    return list(dict.fromkeys(tables))  # dedupe, preserve order


def _split_function_name(name: str) -> list[str]:
    """Split a function name into meaningful words.

    Handles ``snake_case``, ``camelCase``, and ``PascalCase``.
    """
    # Split on underscores first
    parts = name.replace("-", "_").split("_")
    # Then split camelCase/PascalCase within each part
    words: list[str] = []
    for part in parts:
        tokens = re.sub(r"([a-z])([A-Z])", r"\1 \2", part).split()
        words.extend(t.lower() for t in tokens if t)
    return words


def _make_wrong_path(correct: str) -> str:
    """Generate a plausible but incorrect path for a distractor."""
    dirname = os.path.dirname(correct) or "/"
    basename = os.path.basename(correct)
    distractors = [
        os.path.join(dirname, "tmp_" + basename),
        os.path.join(os.path.dirname(dirname) or "/", basename),
        correct + ".bak",
    ]
    return random.choice(distractors)


# ---------------------------------------------------------------------------
# QuizChallenge
# ---------------------------------------------------------------------------


class QuizChallenge:
    """Comprehension quiz for HIGH-risk actions.

    Analyses the :class:`ActionContext` to programmatically generate 1-3
    multiple-choice or fill-in-the-blank questions about the pending action.
    The operator must answer at least one question correctly to pass.

    Parameters
    ----------
    max_questions:
        Upper bound on the number of questions to generate (1-3).
    min_correct:
        Minimum correct answers required to pass.
    min_review_seconds:
        Minimum wall-clock seconds before the first question is shown.
    """

    def __init__(
        self,
        max_questions: int = 3,
        min_correct: int = 1,
        min_review_seconds: float = 10.0,
    ) -> None:
        self.max_questions = max(1, min(max_questions, 3))
        self.min_correct = max(1, min_correct)
        self.min_review_seconds = min_review_seconds

    @property
    def challenge_type(self) -> ChallengeType:
        return ChallengeType.QUIZ

    # -- question generation ----------------------------------------------

    def generate_questions(self, ctx: ActionContext, risk: RiskAssessment) -> list[Question]:
        """Build a list of questions from the action context.

        Strategy priority:
        1. File-path arguments   -> "What directory / file will be affected?"
        2. Numeric arguments     -> "What is the value of <param>?"
        3. SQL-like strings      -> "Which table will be affected?"
        4. Fallback              -> "What will this function do?" (from name)
        """
        flat = _flatten_args(ctx)
        questions: list[Question] = []

        # Strategy 1: file paths
        paths = _find_paths(flat)
        if paths:
            path = paths[0]
            dirname = os.path.dirname(path) or "/"
            basename = os.path.basename(path)
            # Ask about the directory
            wrong_dirs = [
                os.path.dirname(os.path.dirname(path)) or "/",
                "/tmp",
                "/var/log",
            ]
            # Filter out correct answer from distractors and dedupe
            wrong_dirs = list(dict.fromkeys(d for d in wrong_dirs if d != dirname))
            options = [dirname] + wrong_dirs[:3]
            random.shuffle(options)
            questions.append(
                Question(
                    text="What directory will be affected by this action?",
                    correct_answer=dirname,
                    options=options,
                )
            )
            # Optionally ask about the specific file
            if len(questions) < self.max_questions and basename:
                questions.append(
                    Question(
                        text="Which file will this action operate on?",
                        correct_answer=basename,
                        options=None,  # free text
                    )
                )

        # Strategy 2: numeric arguments
        numbers = _find_numbers(flat)
        if numbers and len(questions) < self.max_questions:
            name, num_str = numbers[0]
            label = f"parameter '{name}'" if name else "this action"
            # Create wrong options
            try:
                num_val = float(num_str)
                wrong = [
                    str(int(num_val * 2)) if num_val == int(num_val) else f"{num_val * 2:.2f}",
                    str(int(num_val + 10)) if num_val == int(num_val) else f"{num_val + 10:.2f}",
                    str(max(0, int(num_val - 1))) if num_val == int(num_val) else f"{max(0.0, num_val - 1):.2f}",
                ]
            except ValueError:
                wrong = ["0", "100", "42"]

            wrong = list(dict.fromkeys(w for w in wrong if w != num_str))[:3]
            options = [num_str] + wrong
            random.shuffle(options)
            questions.append(
                Question(
                    text=f"What is the numeric value for {label}?",
                    correct_answer=num_str,
                    options=options,
                )
            )

        # Strategy 3: SQL table names
        tables = _find_sql_tables(flat)
        if tables and len(questions) < self.max_questions:
            table = tables[0]
            wrong_tables = ["users", "logs", "tmp_data", "sessions", "config"]
            wrong_tables = [t for t in wrong_tables if t.lower() != table.lower()][:3]
            options = [table] + wrong_tables
            random.shuffle(options)
            questions.append(
                Question(
                    text="Which database table will be affected?",
                    correct_answer=table,
                    options=options,
                )
            )

        # Strategy 4 (fallback): function-name comprehension
        if not questions:
            words = _split_function_name(ctx.function_name)
            action_verb = words[0] if words else "perform"
            target = " ".join(words[1:]) if len(words) > 1 else "an operation"
            correct = f"{action_verb} {target}"

            wrong = [
                f"read {target}",
                f"list {target}",
                f"validate {target}",
            ]
            wrong = [w for w in wrong if w != correct][:3]
            options = [correct] + wrong
            random.shuffle(options)
            questions.append(
                Question(
                    text=f"What will the function '{ctx.function_name}' do?",
                    correct_answer=correct,
                    options=options,
                )
            )

        return questions[: self.max_questions]

    # -- presentation -----------------------------------------------------

    async def present(self, ctx: ActionContext, risk: RiskAssessment) -> ChallengeResult:
        """Present the quiz to the operator and collect answers."""
        start = time.monotonic()
        loop = asyncio.get_running_loop()

        # -- render action summary ----------------------------------------
        separator = "=" * 60
        print(f"\n{separator}")
        print("  QUIZ CHALLENGE  --  HIGH RISK ACTION")
        print(f"{separator}")
        print(f"  Action: {ctx.function_name}")
        print(f"  Risk:   {risk.level.value.upper()} ({risk.score:.2f})")
        if ctx.function_doc:
            print(f"  Desc:   {ctx.function_doc}")
        print(f"  Call:   {ctx.description}")
        print(f"{separator}")

        # -- enforce minimum review time ----------------------------------
        elapsed = time.monotonic() - start
        if elapsed < self.min_review_seconds:
            remaining = self.min_review_seconds - elapsed
            print(f"  [Review the action for {remaining:.0f}s before answering...]")
            await asyncio.sleep(remaining)

        # -- generate and ask questions -----------------------------------
        questions = self.generate_questions(ctx, risk)
        correct_count = 0

        def _read_input(prompt: str) -> str:
            try:
                return input(prompt).strip()
            except (EOFError, KeyboardInterrupt):
                return ""

        for idx, question in enumerate(questions, start=1):
            print(f"\n  Question {idx}/{len(questions)}: {question.text}")

            if question.options:
                for letter_idx, option in enumerate(question.options):
                    letter = chr(ord("A") + letter_idx)
                    print(f"    {letter}) {option}")
                raw = await loop.run_in_executor(None, lambda: _read_input("  Your answer (letter or value): "))
                # Accept either the letter label or the literal value
                answer = raw
                if len(raw) == 1 and raw.upper().isalpha():
                    choice_idx = ord(raw.upper()) - ord("A")
                    if 0 <= choice_idx < len(question.options):
                        answer = question.options[choice_idx]
            else:
                raw = await loop.run_in_executor(None, lambda: _read_input("  Your answer: "))
                answer = raw

            if answer.lower() == question.correct_answer.lower():
                correct_count += 1
                print("  Correct.")
            else:
                print(f"  Incorrect. (expected: {question.correct_answer})")

        elapsed = time.monotonic() - start
        passed = correct_count >= self.min_correct

        print(f"\n  Result: {correct_count}/{len(questions)} correct -- {'PASSED' if passed else 'FAILED'}")

        return ChallengeResult(
            passed=passed,
            challenge_type=self.challenge_type,
            response_time_seconds=elapsed,
            questions_asked=len(questions),
            questions_correct=correct_count,
            details={
                "questions": [
                    {
                        "text": q.text,
                        "correct_answer": q.correct_answer,
                        "options": q.options,
                    }
                    for q in questions
                ],
            },
        )
